scipy/integrate/quadpack.h

```diff
@@ -39,93 +39,161 @@ the result tuple when the full_output argument is non-zero.
-    quadpack_python_function = (PyObject *)store->global0;
-    quadpack_extra_arguments = (PyObject *)store->global1;
+static void quad_restore_func(QStorage * store, int *ierr)
+{
+    quadpack_python_function = (PyObject *) store->global0;
+    quadpack_extra_arguments = (PyObject *) store->global1;
     memcpy(&quadpack_jmpbuf, &(store->jmp), sizeof(jmp_buf));
     Py_XDECREF(store->arg);
     if (ierr != NULL) {
-        if (PyErr_Occurred()) {
-            *ierr = 80;             /* Python error */
-            PyErr_Clear();
-        }
+      	if (PyErr_Occurred()) {
+      	    *ierr = 80;		/* Python error */
+      	    PyErr_Clear();
+      	}
     }
 }
 
-static int
-init_ctypes_func(QStorage *store, PyObject *fun) {
+static int init_ctypes_func(QStorage * store, PyObject * fun)
+{
     store->global0 = quadpack_ctypes_function;
     store->global1 = get_ctypes_function_pointer(fun);
-    if (store->global1 == NULL) return NPY_FAIL;
+    if (store->global1 == NULL)
+      	return NPY_FAIL;
     quadpack_ctypes_function = store->global1;
     return NPY_SUCCEED;
 }
 
-static void
-restore_ctypes_func(QStorage *store) {
+static void restore_ctypes_func(QStorage * store)
+{
     quadpack_ctypes_function = store->global0;
 }
 
+static double *c_array_from_tuple(PyObject * tuple)
+{
+    /* Accepts Python tuple and converts to double array in c for use in
+     * multivariate ctypes */
+    if (!PyTuple_CheckExact(tuple))
+      	return NULL;		/*Ensure python tuple is passed in */
+    Py_ssize_t n_args = PyTuple_Size(tuple);
+    Py_ssize_t i = 0;
+    double *array = (double *) malloc(sizeof(double) * (n_args + 1));
+    PyObject *item = NULL;
+    array[0] = 0.0;
+    for (i = 0; i < n_args; i++) {
+      	item = PyTuple_GetItem(tuple, i);
+      	array[i + 1] = PyFloat_AsDouble(item);
+    }
+    return array;
+}
 
+static int
+init_c_multivariate(QStorage * store, PyObject * f, PyObject * args)
+{
+    /*Initialize function of n+1 variables
+     * Parameters: 
+     * store - Qstorage pointer to hold current state of stack
+     * f - Pyobject function pointer to function to evaluate
+     * n - integer number of extra parameters 
+     * args - Python tuple with parameters x[1] ... x[n]
+     * Output:
+     * NPY_FAIL on failure 
+     * NPY_SUCCEED on success
+     */
+
+    /*Store current parameters */
+    store->global0 = global_function;
+    store->global1 = global_n_args;
+    store->arg = global_args;
+
+    /*Set new parameters */
+    if ((global_function = get_ctypes_function_pointer(f)) == NULL) {
+      	PyErr_SetString(quadpack_error,
+      			"Ctypes function not correctly initialized");
+      	return NPY_FAIL;
+    }
+    if ((global_args = c_array_from_tuple(args)) == NULL) {
+      	PyErr_SetString(quadpack_error,
+      			"Extra Arguments must be in a tuple");
+      	return NPY_FAIL;
+    }
+    int n_args_ref = PyTuple_Size(args);
+    global_n_args = &n_args_ref;
```

Comment:
This stores a global pointer to a local variable that will soon go out of scope. Clearly the value dereferenced later will be non-sensical.